### ISSUE TITLE
expose sincos?

### ISSUE THREAD

**Issue by mlubin** [JuliaLang/julia#10442]
GNU libm and openlibm both implement `sincos`:

```
void sincos(double x, double *sin, double *cos);


DESCRIPTION
       Several  applications need sine and cosine of the same angle x.  This function
       computes both at the same time, and stores the results in *sin and *cos.
```

LLVM also has a sincos intrinsic.

Could we expose this? I observed a 30% speedup by calling sincos over sin and cos separately when using the openlibm implementation. Ideally this would be a compiler optimization, but since we're not using the LLVM intrinsics for sin and cos, I'm guess that we won't easily get LLVM to do this for us for free.
In DualNumbers, this would be very useful since computing either sin or cos of a `Dual` requires both values, which in turn would speed up derivative evaluation in JuMP.

**Comment by mlubin** [comment]
I would submit a PR for this, but I don't know how to wrap the libm signature of ``sincos`` without allocating a temporary array for the output.

**Comment by simonbyrne** [comment]
Some previous discussion on julia-users:
https://groups.google.com/d/topic/julia-users/368fpIGYRrY/discussion

I think the easiest solution might be simply to implement these in Julia: we already have the `sin`/`cos` "kernel" functions, as they are used in `sinpi`/`cospi`, so all that's really needed is the range-reduction code (which would be useful for #8799 anyway). I have made a start here:
https://github.com/simonbyrne/libm.jl/blob/master/src/rempio2.jl
It still requires an implementation of Payne-Hanek reduction (required for arguments > `1.6e6`) but until that gets done, we could just call the existing C code.

**Comment by mlubin** [comment]
@simonbyrne, could you give a hint on how to put all the pieces together to test this out if I don't care about arguments ``> 1.6e6``?

**Comment by simonster** [comment]
We should also use this for `cis`, which it looks like currently calls `sin` and `cos` separately.

**Comment by simonbyrne** [comment]
@mlubin I had something in mind like this:
https://github.com/JuliaLang/julia/compare/master...simonbyrne:sincos

It still does seem to be a bit slower, but I haven't really done any profiling to find out where the problem is.

**Comment by mlubin** [comment]
Cool! Is there any extra speedup for combining the
```
    z = x.hi*x.hi
    w = z*z
```
computations across ``cos_kernel`` and ``sin_kernel``?

**Comment by simonbyrne** [comment]
From what I can tell, if you `@inline` the kernel functions, then LLVM is able to combine them itself.

**Comment by vtjnash** [comment]
> I would submit a PR for this, but I don't know how to wrap the libm signature of sincos without allocating a temporary array for the output.

with the ccall work that's being merged soon, you can replace 1-value Arrays with `Ref` objects. eventually, those should even get allocated on the stack too.

**Comment by zsoerenm** [comment]
This can also give a tremendous speedup for `exp(1im * arg)`
as it is the same as `complex(cos(arg), sin(arg))`
Further details are discussed here: https://discourse.julialang.org/t/why-is-this-julia-code-considerably-slower-than-matlab/2365/45

**Comment by yuyichao** [comment]
This requires better optimization before it can be implemented.

**Comment by yuyichao** [comment]
What's the "other stuff is suggested"?

**Comment by StefanKarpinski** [comment]
I.e. why can't we expose a `sincos` function before optimizing it fully? Then we can work on optimization at any point we feel like it?

**Comment by yuyichao** [comment]
We can define `sincos`, which we kind of do already and it's called `cis`. We can't expose `sincos` from libm since that'll be slower.

**Comment by StefanKarpinski** [comment]
Well, `cis` kind of computes `sincos` if you understand complex exponentials. I suppose if we allowed destructuring of complex numbers then one could just write this: `s, c = cis(x)`. Actually, that would be fairly handy for rationals as well.

**Comment by simonbyrne** [comment]
meh: destructuring complex numbers seems a little too magical. I don't think it's too hard to do `reim(cis(x))`. The other issue is that it gives them in the opposite order.

I would be in favour of a `sincos` function (as well as `sincospi` and possibly a `sincosd` as well).

**Comment by Datseris** [comment]
Hi,

sorry for posting on such an old issue, but I also need the function `sincos` for a library I am using. However, even though this issue thread seems to report that the function is accessible (and in fact it is in [`julia/exports.jl`](https://github.com/JuliaLang/julia/blob/71fbc70ef9c3d4a7fee24d9835a42034170be97f/base/exports.jl#L331)), for some reason I don't have access to it:

```
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: https://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.6.2 (2017-12-13 18:08 UTC)
 _/ |\__'_|_|_|\__'_|  |  Official http://julialang.org/ release
|__/                   |  x86_64-w64-mingw32

julia> sincos
ERROR: UndefVarError: sincos not defined

help?> sincos
search:

Couldn't find sincos
Perhaps you meant sinc, sin, sind, sinh, sinpi, sizeof, cos, acos, minabs, info or asin
  No documentation found.

  Binding sincos does not exist.
```

I am on Windows 10 64-bit. Can you tell me where I can find the function?

**Comment by fredrikekre** [comment]
> Can you tell me where I can find the function?

https://github.com/JuliaLang/julia/pull/21589 was merged post 0.6. It's on julia master.
